include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -878,6 +879,47 @@ struct __brick_shift_left
     }
 };
 
+struct __brick_assign
+{
+    template <typename _T1, typename _T2>
+    void
+    operator()(const _T1& __a, _T2&& __b) const
+    {
+        ::std::get<0>(::std::forward<_T2>(__b)) = ::std::get<2>(__a);     //key
+        ::std::get<1>(::std::forward<_T2>(__b)) = ::std::get<0>(__a) + 1; //index
+    }
+};
+
+template <typename _BinaryOperator>
+struct __brick_reduce_idx
+{
+    __brick_reduce_idx(const _BinaryOperator& __b) : __binary_op(__b) {}
+
+    template <typename _Idx, typename _Values>
+    auto
+    reduce(_Idx __i, _Idx __j, const _Values& __values) const
+    {
+        auto __res = __values[__i];
+        for (++__i; __i < __j; ++__i)
+            __res = __binary_op(__res, __values[__i]);
+
+        return __res;
+    }
+
+    template <typename _ItemId, typename _ReduceIdx, typename _Values, typename _OutValues>
+    void
+    operator()(const _ItemId __idx, const _ReduceIdx& __reduce_idx, const _Values& __values,
+               _OutValues& __out_values) const
+    {
+        using __value_type = decltype(__reduce_idx[__idx]);
+        __value_type __left_operand = (__idx == 0) ? __value_type(0) : __reduce_idx[__idx - 1];
+        __out_values[__idx] = reduce(__left_operand, __reduce_idx[__idx], __values);
+    }
```

Comment:
I am confused by `__reduce_idx[__idx]`; and `__left_operand` also is unclear. Please think of better naming that explains the behavior.

Reply:
New names have been introduced. `__left_operand` is now `__segment_begin` and `__reduce_idx` is `__segment_ends` to document that what it holds is the end of each segment to be reduced.